File: src/VoiceGenerator.cpp
#include "VoiceGenerator.h"
#include "AUI/Logging/ALogger.h"
#include "AUI/IO/AFileOutputStream.h"
#include <chrono>

static constexpr auto LOG_TAG = "VoiceGenerator";

AFuture<VoiceGenerator::VoiceMessage> VoiceGenerator::generate(AString text, AString languageCode, double speed) {
    ALogger::info(LOG_TAG) << "Generating voice message for text: " << text.substr(0, 50) << "...";

    try {
        APath voiceDir("data/voice_messages");
        voiceDir.makeDirs();

        ElevenLabsClient ttsClient{
            .baseUrl = "https://api.elevenlabs.io/",
            .apiKey = mApiKey,
            .voiceId = mVoiceId,
        };

        ElevenLabsClient::TextToSpeechRequest request{
            .text = text,
            .model_id = "eleven_v3",
            .language_code = languageCode,
            .voice_settings = {.speed = speed},
        };

        auto ttsResponse = co_await ttsClient.textToSpeech(request);

        if (ttsResponse.audioData.empty()) {
            throw AException("ElevenLabs returned empty audio data");
        }

        auto timestamp = std::chrono::system_clock::now().time_since_epoch().count();
        APath outputPath = voiceDir / "{}.mp3"_format(timestamp);

        AFileOutputStream stream(outputPath);
        stream.write(
            reinterpret_cast<const char*>(ttsResponse.audioData.data()),
            ttsResponse.audioData.getSize()
        );
        stream.close();

        ALogger::info(LOG_TAG) << "Voice message saved to: " << outputPath.absolute();

        co_return VoiceMessage{ .path = outputPath.absolute() };
    } catch (const AException& e) {
        ALogger::err(LOG_TAG) << "Failed to generate voice message: " << e;
        throw;
    }
}

File: src/VoiceGenerator.py
import asyncio
from pathlib import Path
from typing import Optional
from dataclasses import dataclass


@dataclass
class VoiceMessage:
    """Generated voice message"""
    path: str


class VoiceGenerator:
    """Voice generation using ElevenLabs API"""
    
    def __init__(self, api_key: str, voice_id: str = "pPdl9cQBQq4p6mRkZy2Z"):
        self.api_key = api_key
        self.voice_id = voice_id
    
    async def generate(self, text: str, language_code: str = "en", speed: float = 1.0) -> VoiceMessage:
        """Generate voice message"""
        import aiohttp
        
        url = "https://api.elevenlabs.io/v1/text-to-speech"
        
        async with aiohttp.ClientSession() as session:
            async with session.post(
                url,
                headers={
                    "xi-api-key": self.api_key,
                    "Content-Type": "application/json"
                },
                json={
                    "text": text,
                    "model_id": "eleven_multilingual_v2",
                    "language_code": language_code,
                    "voice_settings": {
                        "speed": speed
                    }
                }
            ) as response:
                if response.status == 200:
                    audio_data = await response.read()
                    
                    # Save to file
                    voice_dir = Path("data/voice_messages")
                    voice_dir.mkdir(parents=True, exist_ok=True)
                    
                    timestamp = asyncio.get_event_loop().time()
                    output_path = voice_dir / f"{timestamp}.mp3"
                    
                    with open(output_path, 'wb') as f:
                        f.write(audio_data)
                    
                    return VoiceMessage(path=str(output_path))
                else:
                    raise Exception(f"Failed to generate audio: {response.status}")